File: code_snippets/ts/src/develop/awakeable.ts
import * as restate from "@restatedev/restate-sdk";

const service = restate.service({
  name: "Awakeable",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      // <start_here>
      // <mark_1>
      const { id, promise } = ctx.awakeable<string>();
      // </mark_1>

      // <mark_2>
      await ctx.run(() => triggerTaskAndDeliverId(id));
      // </mark_2>

      // <mark_3>
      const payload = await promise;
      // </mark_3>
      // <end_here>

      // <start_resolve>
      ctx.resolveAwakeable(id, "hello");
      // <end_resolve>

      // <start_reject>
      ctx.rejectAwakeable(id, "my error reason");
      // <end_reject>
    },
  },
});

function triggerTaskAndDeliverId(awakeableId: string) {
  return "123";
}

File: code_snippets/go/develop/awakeable.go
package develop

import (
	"fmt"

	restate "github.com/restatedev/sdk-go"
)

type Awakeable struct{}

func (Awakeable) Greet(ctx restate.Context, name string) error {
	// <start_here>
	// <mark_1>
	awakeable := restate.Awakeable[string](ctx)
	awakeableId := awakeable.Id()
	// </mark_1>

	// <mark_2>
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return triggerTaskAndDeliverId(awakeableId)
	}); err != nil {
		return err
	}
	// </mark_2>

	// <mark_3>
	payload, err := awakeable.Result()
	if err != nil {
		return err
	}
	// </mark_3>
	// <end_here>

	_ = payload

	// <start_resolve>
	restate.ResolveAwakeable(ctx, awakeableId, "hello")
	// <end_resolve>

	// <start_reject>
	restate.RejectAwakeable(ctx, awakeableId, fmt.Errorf("my error reason"))
	// <end_reject>

	return nil
}

func triggerTaskAndDeliverId(awakeableId string) (string, error) {
	return "123", nil
}
